keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -90,23 +90,52 @@ We propose to
+We do this in the serializer by passing a real defaulter to [`versioningserializer.NewCodec`](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/serializer/versioning/versioning.go#L49) such that defaulting is done natively just after the binary payload has been unmarshalled into an `map[string]interface{}` and pruning of [KEP: Pruning for CustomResources](/keps/sig-api-machinery/20180731-crd-pruning.md) was done.
+
+Like for native resources, we do defaulting
+
+* during request payload deserialization
+* after mutating webhook admission
+* during read from storage.
+     
+Note: like for native resources, we do not default after webhook conversions. Hence, webhook conversions must be complete in the sense that they return defaulted objects. Technically we could do defaulting, but to match native resources, we do not.
+     
+Compare the yellow boxes in the following figure:
 
 ![Decoding steps which must apply defaults](20190426-crd-defaulting-pipeline.png)
 
-We rely on the validation steps in the request pipeline to verify that the default value is of the right type.
+We rely on the validation steps in the request pipeline to verify that the default value validates value validation. We will check the types using the _structural schema_ during CRD creation and update though.
 
-The `default` field in the CRD types is considered alpha quality. We will add a `CustomResourceDefaulting` feature gate. Values for `default` will be rejected if the gate is not enabled. 
+The `default` field in the CRD types is considered alpha quality. We will add a `CustomResourceDefaulting` feature gate. Values for `default` will be rejected if the gate is not enabled and there have not been `default` values set before (ratcheting validation). 
 
-[Kubebuilder's crd-gen](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/crd) can make use of this feature by adding another tag, e.g. `// +default=<arbitrary-json-value>`. Defaults are arbitrary JSON values, which must also validate and are not subject to pruning (defaulting happens after pruning). This is an implicit assumption that will be checked by the apiserver.
+[Kubebuilder's crd-gen](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/crd) can make use of this feature by adding another tag, e.g. `// +default=<arbitrary-json-value>`. Defaults are arbitrary JSON values, which must also validate (types are checked during CRD creation and update, value validation is checked for requests, but not for etcd reads) and are not subject to pruning (defaulting happens after pruning).
```

Comment:
is it possible to ensure the default value does not contain a field that would get dropped via pruning? this would help prevent typos. I'm willing to do more expensive things in CRD update validation to improve the user experience, as long as it falls out in a relatively straightforward way in the code

Reply:
for a given `JSONSchemaProps` object, I was envisioning checks like this:

```go
reflect.DeepEqual(
  props.Default,
  prune(
    props.Default,
    makeStructuralSchema(props),
    isPreservingUnknownFields, /* from parent schemas or CRD field */
  ),
)

validate(props.Default, ConvertJSONSchemaProps(props))
```

Reply:
yeah, we can do that.

Reply:
added